### PR TITLE
Feature/covid geocoder proximity

### PR DESCRIPTION
**Pull request by adil-uddin**
https://docs.google.com/document/d/1CSN5eZKQMetYCOyDTnAPuzIXZ6vcZbTj5gX3OUMNgAk/edit

##### SUMMARY

Added the `Default geocoder location` field in project settings. Will be used in setting the proximity of Mapbox geocoder for different projects.

### CODE REVIEW

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 1]
```diff
@@ -1,5 +1,13 @@
 /* globals Formplayer, EntrySingleAnswer, moment */
 
+/**
+ * Import vars sent from server
+ */
+var initialPageData = hqImport("hqwebapp/js/initial_page_data").get;
+var MAPBOX_ACCESS_TOKEN = initialPageData("mapbox_access_token");
```
[the remark this answers is not in the document]
Comment (adil-uddin) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js:7]: @orangejenny suggestion please

**Comment by adil-uddin** [comment]
@orangejenny Kat mentioned in the spec that if the search bar is empty we should show `Boston` as default in the search bar. But I'm afraid this might create an illusion to user that the location is already set and the user might not click on submit button. I'm leaving it empty for now.

**Comment by orangejenny** [comment]
@knguyen142 and @calellowitz might also have feedback

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 2]
```diff
@@ -305,13 +313,13 @@ function AddressEntry(question, options) {
...
         var geocoder = new MapboxGeocoder({
-            accessToken: window.MAPBOX_ACCESS_TOKEN,
-            types: 'address',
+            accessToken: MAPBOX_ACCESS_TOKEN,
+            types: "address",
             enableEventLogging: false,
-            // proximity set to NYC
-            proximity: { longitude: -74.006058, latitude: 40.712772},
-            getItemValue: self.geocoderItemCallback,
+            proximity: defaultGeocoderLocation.coordinates,
```
Comment (knguyen142) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js:322]: You can also do a check if `defaultGeocoderLocation.coordinates` and call `setProximity` https://github.com/mapbox/mapbox-gl-geocoder/blob/master/API.md#setproximity. I'm assuming an undefined proximity does not have negative sideeffects? What does it default to?

File: corehq/apps/cloudcare/static/cloudcare/js/preview_app/main.js [hunk 3]
```diff
@@ -28,4 +28,4 @@ hqDefine("cloudcare/js/preview_app/main", function () {
             $(sc).addClass('has-scrollbar-' + scrollWidth);
         });
     });
-});
+});
```
Comment (knguyen142) [inline corehq/apps/cloudcare/static/cloudcare/js/preview_app/main.js:31]: nit: but can you add a new line?

File: corehq/apps/domain/static/domain/js/info_basic.js [hunk 4]
```diff
@@ -1,31 +1,72 @@
...
+        "@mapbox/mapbox-gl-geocoder/dist/mapbox-gl-geocoder.min",
+        "hqwebapp/js/initial_page_data",
+        "hqwebapp/js/select_2_ajax_widget", // for call center case owner
+        "select2/dist/js/select2.full.min",
+    ],
+    function ($, _, MapboxGeocoder,initialPageData) {
+        $(function () {
+            // var initialPageData = hqImport("hqwebapp/js/initial_page_data").get;
+            var MAPBOX_ACCESS_TOKEN = initialPageData.get("mapbox_access_token");
+            console.log(MAPBOX_ACCESS_TOKEN);
```
Comment (knguyen142) [inline corehq/apps/domain/static/domain/js/info_basic.js:16]: Debugging?

File: corehq/apps/reports_core/templates/reports_core/base_template_new.html [hunk 5]
```diff
@@ -57,6 +57,7 @@
   {% initial_page_data 'table_default_rows' report_table.default_rows|default:10 %}
   {% initial_page_data 'report_slug' report.slug %}
   {% initial_page_data 'MAPBOX_ACCESS_TOKEN' MAPBOX_ACCESS_TOKEN %}
+  {% initial_page_data 'default_geocoder_location' default_geocoder_location %}
```
Comment (knguyen142) [inline corehq/apps/reports_core/templates/reports_core/base_template_new.html:60]: Does the reports app need this variable?

File: corehq/apps/domain/static/domain/js/info_basic.js [hunk 6]
```diff
@@ -1,31 +1,72 @@
...
+    "domain/js/info_basic",
+    [
+        "jquery",
+        "underscore",
+        "@mapbox/mapbox-gl-geocoder/dist/mapbox-gl-geocoder.min",
+        "hqwebapp/js/initial_page_data",
+        "hqwebapp/js/select_2_ajax_widget", // for call center case owner
+        "select2/dist/js/select2.full.min",
+    ],
+    function ($, _, MapboxGeocoder,initialPageData) {
```
Comment (knguyen142) [inline corehq/apps/domain/static/domain/js/info_basic.js:12]: I'll need to look at this more because I'm not familiar with this part of the codebase. But can you take screenshots and/or link to it deployed in staging so I can see what it's supposed to be like?

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js (hunk 2 above)
Comment (adil-uddin) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: docs don't say much about undefined proximity, so I did what you suggested.

File: corehq/apps/domain/static/domain/js/info_basic.js (hunk 6 above)
Comment (adil-uddin) [reply corehq/apps/domain/static/domain/js/info_basic.js]: Screenshot from 2020-08-26 16-41-44

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js:7]: This is fine for now. This file doesn't use our [standard module patterns](https://github.com/dimagi/commcare-hq/blob/master/docs/js-guide/code-organization.md), which isn't great, but it's out of scope to address in this PR.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 7]
```diff
@@ -305,14 +313,15 @@ function AddressEntry(question, options) {
...
+        var defaultGeocoderLocation = initialPageData('default_geocoder_location') || {};
         var geocoder = new MapboxGeocoder({
-            accessToken: window.MAPBOX_ACCESS_TOKEN,
-            types: 'address',
+            accessToken: MAPBOX_ACCESS_TOKEN,
+            types: "address",
             enableEventLogging: false,
-            // proximity set to NYC
-            proximity: { longitude: -74.006058, latitude: 40.712772},
-            getItemValue: self.geocoderItemCallback,
```
Comment (orangejenny) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js:314]: Why was this line removed?

File: corehq/apps/cloudcare/views.py [hunk 8]
```diff
@@ -187,6 +189,9 @@ def get_option_apps(self, request, domain):
         apps = self.get_web_apps_available_to_user(domain, restore_as)
         return JsonResponse(apps, safe=False)
 
+    def _get_domain_obj(self, domain):
```
Comment (orangejenny) [inline corehq/apps/cloudcare/views.py:192]: Minor: you don't really gain anything by making this a function, given that it's only called once and it's a one-liner.

File: corehq/apps/cloudcare/views.py [hunk 9]
```diff
@@ -266,9 +277,11 @@ def _default_lang():
         # default language to user's preference, followed by
         # first app's default, followed by english
         language = request.couch_user.language or _default_lang()
+        domainobj = self._get_domain_obj(domain)
```
Comment (orangejenny) [inline corehq/apps/cloudcare/views.py:280]: Naming conventions: `domainobj` => `domain_obj` - although the same comment as above applies, I'd just update the line below to be `"default_geocoder_location": domainobj.Domain.get_by_name(domain)`.

File: corehq/apps/domain/forms.py [hunk 10]
```diff
@@ -315,6 +315,13 @@ class DomainGlobalSettingsForm(forms.Form):
     )
     default_timezone = TimeZoneChoiceField(label=ugettext_noop("Default Timezone"), initial="UTC")
 
+    default_geocoder_location = Field(
+        widget=GeoCoderInput(attrs={'placeholder': ugettext_lazy('Select a location')}),
+        label=ugettext_noop("Default geocoder location"),
+        required=False,
+        help_text=ugettext_lazy("Please select the default location for the geocoder used in your project space.")
```
Comment (orangejenny) [inline corehq/apps/domain/forms.py:322]: As we discussed, I think this wording isn't totally clear. I left a comment on the spec to get input from Kat.

File: corehq/apps/domain/models.py [hunk 11]
```diff
@@ -309,6 +309,7 @@ class Domain(QuickCachedDocumentMixin, BlobMixin, Document, SnapshotMixin):
     is_active = BooleanProperty()
     date_created = DateTimeProperty()
     default_timezone = StringProperty(default=getattr(settings, "TIME_ZONE", "UTC"))
+    default_geocoder_location = StringProperty()
```
Comment (orangejenny) [inline corehq/apps/domain/models.py:312]: There's a `DictProperty` you can use instead of storing this as a string, which would make it more semantic.

File: corehq/apps/domain/templates/domain/admin/info_basic.html [hunk 12]
```diff
@@ -2,8 +2,11 @@
 {% load crispy_forms_tags %}
 {% load hq_shared_tags %}
 
+
 {% requirejs_main 'domain/js/info_basic' %}
 
 {% block page_content %}
+  <link rel="stylesheet" href="{% static '@mapbox/mapbox-gl-geocoder/dist/mapbox-gl-geocoder.css' %}" type="text/css" />
```
Comment (orangejenny) [inline corehq/apps/domain/templates/domain/admin/info_basic.html:9]: There's a stylesheets block you can put this in instead, so it gets included with other css [here](https://github.com/dimagi/commcare-hq/blob/92656804df7513dc95826572d7c270855fbf293f/corehq/apps/hqwebapp/templates/hqwebapp/base.html#L56).

File: corehq/apps/hqwebapp/widgets.py [hunk 13]
```diff
@@ -158,3 +158,16 @@ def render(self, name, value, attrs=None, renderer=None):
...
+class GeoCoderInput(Input):
+
+    def __init__(self, attrs=None):
+        super(GeoCoderInput, self).__init__(attrs=attrs)
+
+    def render(self, name, value, attrs=None, renderer=None):
+        self.input_type = 'hidden'
+        output = super(GeoCoderInput, self).render(name, value, attrs, renderer)
+        return mark_safe("""
+            <div id="geocoder-proximity">{}</div>
```
Comment (orangejenny) [inline corehq/apps/hqwebapp/widgets.py:172]: This is something of a nitpick, but it'd be better for this to be a class and for the js to loop over all elements with that class, rather than only allowing one geocoder on the page.

File: corehq/apps/reports/static/reports/js/edit_submission.js [hunk 14]
```diff
@@ -2,6 +2,7 @@ hqDefine("reports/js/edit_submission", function () {
     $(function () {
         var initialPageData = hqImport("hqwebapp/js/initial_page_data").get;
         window.MAPBOX_ACCESS_TOKEN = initialPageData('mapbox_access_token'); // maps api is a global variable depended on by touchforms
+        window.DEFAULT_GEOCODER_LOCATION = initialPageData('default_geocoder_location');
```
Comment (orangejenny) [inline corehq/apps/reports/static/reports/js/edit_submission.js:5]: I don't think you need this, and also think you can remove the line above since you got rid of the `window`-related logic.

**Review by orangejenny — COMMENTED** [review]
This looks good. I left a bunch of feedback, but it's all minor stuff.

File: corehq/apps/domain/models.py (hunk 11 above)
Comment (adil-uddin) [reply corehq/apps/domain/models.py]: changed to DictProperty

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js (hunk 7 above)
Comment (adil-uddin) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js:314]: Sorry :( adding back

File: corehq/apps/cloudcare/views.py (hunk 8 above)
Comment (adil-uddin) [reply corehq/apps/cloudcare/views.py]: done

File: corehq/apps/hqwebapp/widgets.py (hunk 13 above)
Comment (adil-uddin) [reply corehq/apps/hqwebapp/widgets.py:172]: exploring on this

File: corehq/apps/domain/forms.py [hunk 15]
```diff
@@ -498,7 +498,7 @@ def save(self, request, domain):
         domain.hr_name = self.cleaned_data['hr_name']
         domain.project_description = self.cleaned_data['project_description']
         domain.default_mobile_ucr_sync_interval = self.cleaned_data.get('mobile_ucr_sync_interval', None)
-        domain.default_geocoder_location = self.cleaned_data.get('default_geocoder_location')
+        domain.default_geocoder_location = json.loads(self.cleaned_data.get('default_geocoder_location','{}'))
```
Comment (orangejenny) [inline corehq/apps/domain/forms.py:501]: Nitpick: I'd do this in the `clean_default_geocoder_location` method. That way if the value is used in multiple places, the json decode only needs to be done once.

File: corehq/apps/hqwebapp/templates/hqwebapp/base.html [hunk 16]
```diff
@@ -53,7 +53,10 @@
       <link rel="apple-touch-icon" sizes="180x180" href="{% static 'hqstyle/images/commcare-icon-180x180.png' %}"/>
     {% endif %}
 
-    {% block stylesheets %}{% endblock stylesheets %}
+    {% block stylesheets %}
```
Comment (orangejenny) [inline corehq/apps/hqwebapp/templates/hqwebapp/base.html:56]: Hey, I think my earlier comment misdirected you. Rather than change this file, in order to make the change only affect the info basic page, you'd add code like this to `info_basic.html`:
```
{% block stylesheets %}{{ block.super }}
   <link rel="stylesheet" href="{% static '@mapbox/mapbox-gl-geocoder/dist/mapbox-gl-geocoder.css' %}" type="text/css" />
{% endblock stylesheets %}
```

**Comment by orangejenny** [comment]
Two more minor comments, then this should be good to go.

File: corehq/apps/reports/static/reports/js/edit_submission.js [hunk 17]
```diff
@@ -1,7 +1,6 @@
 hqDefine("reports/js/edit_submission", function () {
     $(function () {
         var initialPageData = hqImport("hqwebapp/js/initial_page_data").get;
-        window.MAPBOX_ACCESS_TOKEN = initialPageData('mapbox_access_token'); // maps api is a global variable depended on by touchforms
```
Comment (knguyen142) [inline corehq/apps/reports/static/reports/js/edit_submission.js:4]: I know the mapbox api token is used somewhere in reports, are you sure removing this doesn't affect it?

File: corehq/apps/reports/templates/reports/form/edit_submission.html [hunk 18]
```diff
@@ -37,6 +37,7 @@
   {% initial_page_data 'edit_context' edit_context %}
   {% initial_page_data 'edit_formplayer' edit_formplayer %}
   {% initial_page_data 'mapbox_access_token' mapbox_access_token %}
+  {% initial_page_data 'default_geocoder_location' default_geocoder_location %}
```
Comment (knguyen142) [inline corehq/apps/reports/templates/reports/form/edit_submission.html:40]: Also this, is this needed in reports?

**Comment by knguyen142** [comment]
After this is merged, can you update https://docs.google.com/document/d/1M2alDemPuWRaJptTH_XPO-JE9lh49McCpPSjaiI-k54/edit#.  Just a small section on how to change the proximity with screenshots would be good.

File: corehq/apps/reports/static/reports/js/edit_submission.js (hunk 17 above)
Comment (orangejenny) [reply corehq/apps/reports/static/reports/js/edit_submission.js:4]: This isn't a report (despite the name). It's the "edit forms" feature, which lets you take an existing form submission and update it by loading that form into the web apps UI. Ultimately this page includes all of web apps's javascript.

File: corehq/apps/reports/templates/reports/form/edit_submission.html (hunk 18 above)
Comment (orangejenny) [reply corehq/apps/reports/templates/reports/form/edit_submission.html:40]: (same comment as above, this is edit forms)

File: corehq/apps/domain/forms.py [hunk 19]
```diff
@@ -498,7 +504,7 @@ def save(self, request, domain):
         domain.hr_name = self.cleaned_data['hr_name']
         domain.project_description = self.cleaned_data['project_description']
         domain.default_mobile_ucr_sync_interval = self.cleaned_data.get('mobile_ucr_sync_interval', None)
-        domain.default_geocoder_location = json.loads(self.cleaned_data.get('default_geocoder_location','{}'))
+        domain.default_geocoder_location = self.clean_default_geocoder_location()
```
Comment (orangejenny) [inline corehq/apps/domain/forms.py:507]: `clean_default_geocoder_location` gets called automatically and will populate `cleaned_data` with the clean value, the dict. So this can just be `self.cleaned_data['default_geocoder_location']`

Comment (adil-uddin) [reply corehq/apps/domain/forms.py]: Ah, was wondering why are other clean methods not being called anywhere. Thanks Jenny. Made the change.

**Review by orangejenny — APPROVED** [review]
This looks good, thank you for the updates!

I promise most of the javascript in HQ will not turn up **quite** so many lint errors...Web apps is a bit of an odd duck. We're working on standardizing it.

**Comment by orangejenny** [comment]
There have been issues with test failures across HQ today, so I restarted the build, let's see if it passes this time around.

**Comment by orangejenny** [comment]
Test failures all look unrelated. The ICDS team has been working on fixing those issues. Can you merge master and then see if tests pass?

**Comment by orangejenny** [comment]
Tests passed on https://github.com/dimagi/commcare-hq/pull/28468, which just cherry-picks all the commits from this PR onto a new branch, since tests aren't currently working on forks - https://github.com/dimagi/commcare-hq/pull/28379 should fix that. Merging.
